### ISSUE TITLE
Feature request: `jacrev_with_value`

### ISSUE THREAD

**Issue by Linux-cpp-lisp** [pytorch/functorch#333]
The [`grad_with_value`](https://pytorch.org/functorch/generated/functorch.grad_and_value.html) transformation helpfully lets you do a gradient transformation while keeping around arbitrary objects from the forward pass using `has_aux`. This is very useful for potential models in physics, for example, where we are interested both in various energies (forward pass) and associated forces/other gradients of them (backwards pass).

I'm proposing a similar `jacrev_with_value` that has the same options to pass through arbitrary info from the forward pass. This seems like it should be a relatively simple feature, since it doesn't involve any new computations.

**Comment by zou3519** [comment]
Agreed, we should have this

**Comment by zou3519** [comment]
Hi @Linux-cpp-lisp, do you think we should have both (a) `jacrev` API that has a `has_aux` flag and (b) a `jacrev_with_value` API that has a `has_aux` flag, or is just (a) sufficient?

When one has a has_aux flag, they can stash the output of the function in the auxiliary data, something like:
```
def f(x):
  output = x.sin()
  return output, output

jacobian, aux = jacrev(f)(torch.randn(3), has_aux=True)
```
But I imagine it's a little difficult to understand how the `has_aux` flag works when someone first comes across the jacrev API and something named `jacrev_and_value` would obviously return the jacobian of f and the output of f

**Comment by zou3519** [comment]
> The aux can be any object (so some big list/dict of tensors), right?

Yes, the aux can be any object. Cool, it sounds like adding a `has_aux` to jacrev would be sufficient for your use case. I'll try to do that soon (today or tomorrow)

**Comment by Linux-cpp-lisp** [comment]
Great, thanks @zou3519 !

**Comment by zou3519** [comment]
Added in https://github.com/pytorch/functorch/pull/335, please let me know if that works
